yarn-project/acir-simulator/src/avm/avm_context.ts

```diff
@@ -1,140 +1,179 @@
+import { AvmExecutionError, InvalidProgramCounterError, NoBytecodeForContractError } from './errors.js';
+import { initExecutionEnvironment, initInitialMachineState } from './fixtures/index.js';
+import { AvmWorldStateJournal } from './journal/journal.js';
+import { type Instruction } from './opcodes/index.js';
 import { decodeFromBytecode } from './serialization/bytecode_serialization.js';
 
-// FIXME: dependency cycle.
-
 /**
- * Avm Executor manages the execution of the AVM
- *
- * It stores a state manager
+ * Avm Context manages the state and execution of the AVM
  */
 export class AvmContext {
   /** Contains constant variables provided by the kernel */
-  private executionEnvironment: AvmExecutionEnvironment;
+  public environment: AvmExecutionEnvironment = initExecutionEnvironment();
+  /** VM state that is modified on an instruction-by-instruction basis */
+  public machineState: AvmMachineState;
   /** Manages mutable state during execution - (caching, fetching) */
-  private journal: AvmJournal;
+  public worldState: AvmWorldStateJournal;
+
+  /** The public contract code corresponding to this context's contract class */
+  private instructions: Instruction[];
 
-  constructor(executionEnvironment: AvmExecutionEnvironment, journal: AvmJournal) {
-    this.executionEnvironment = executionEnvironment;
-    this.journal = journal;
+  /** Stage data for public kernel (1-kernel-per-call).
+   *  Shouldn't be necessary once kernel processes an entire AVM Session. */
+  // TODO(4293): Integrate simulator with public kernel
+  //private nestedExecutions: PublicExecutionResult[] = [];
+
+  constructor(
+    worldState: AvmWorldStateJournal,
+    environment: AvmExecutionEnvironment = initExecutionEnvironment(),
+    /** Some initial values for machine state */
+    initialMachineState: InitialAvmMachineState = initInitialMachineState(),
+    private log = createDebugLogger('aztec:avm_simulator:avm_context'),
+  ) {
+    this.worldState = worldState;
+    this.environment = environment;
+    this.machineState = new AvmMachineState(initialMachineState);
+    // Bytecode is fetched and instructions are decoded in async init()
+    this.instructions = [];
   }
 
   /**
-   * Call a contract with the given calldata
-   *
-   * - We get the contract from storage
-   * - We interpret the bytecode
-   * - We run the interpreter
-   *
+   * Set instructions directly (execute() will then skip bytecode fetch & decode)
+   * Warning: FOR TESTING PURPOSES ONLY!
+   * @param instructions - The decoded instructions to inject into this context
    */
-  async call(): Promise<AvmMessageCallResult> {
-    // NOTE: the following is mocked as getPublicBytecode does not exist yet
-    const selector = new FunctionSelector(0);
-    const bytecode = await this.journal.hostStorage.contractsDb.getBytecode(
-      this.executionEnvironment.address,
-      selector,
-    );
+  setInstructions(instructions: Instruction[]) {
```

Comment:
This feels strange. Consider having an `execute()` that calls another (public) `execute(Instruction[])` instead.